charmcraft/jujuignore.py

```diff
@@ -0,0 +1,235 @@
+    return rule
+
+
+def _rule_to_regex(rule):
+    """Turn a rule into a regex that we can use.
+
+    This assumes that all the meta processing, like 'ends with /' and 'starts with !' have
+    already been checked.
+    """
+    # Things we currently care about:
+    # * = matches only within a directory "[^/]*"
+    # ** = matches across directories ".*"
+    # ? = matches a single character
+    # [0-9] can match anything 0-9
+    # This is taken from fnmatch.fnmatch, but that doesn't handle '**' and '*' also matches
+    # directories
+
+    i, n = 0, len(rule)
+    res = ''
+    while i < n:
+        c = rule[i]
+        i += 1
+        if c == '*':
+            if i < n and rule[i] == '*':
+                i += 1
+                res += '.*'
+            else:
+                res += '[^/]*'
+        elif c == '?':
+            res += '.'
+        elif c == '[':
+            j = i
+            if j < n and rule[j] == '!':
+                j += 1
+            if j < n and rule[j] == ']':
+                j += 1
+            while j < n and rule[j] != ']':
+                j += 1
+            if j >= n:
+                res += '\\['
+            else:
+                stuff = rule[i:j]
+                if '--' not in stuff:
+                    stuff = stuff.replace('\\', r'\\')
+                else:
+                    chunks = []
+                    k = i + 2 if rule[i] == '!' else i + 1
+                    while True:
+                        k = rule.find('-', k, j)
+                        if k < 0:
+                            break
+                        chunks.append(rule[i:k])
+                        i = k + 1
+                        k += 3
+                    chunks.append(rule[i:j])
+                    # Escape backslashes and hyphens for set difference (--).
+                    # Hyphens that create ranges shouldn't be escaped.
+                    stuff = '-'.join(s.translate({
+                        ord('\\'): r'\\',
+                        ord('-'): r'\-',
+                    }) for s in chunks)
+                # Escape set operations (&&, ~~ and ||).
+                stuff = re.sub(r'([&~|])', r'\\\1', stuff)
+                i = j + 1
+                if stuff[0] == '!':
+                    stuff = '^' + stuff[1:]
+                elif stuff[0] in ('^', '['):
+                    stuff = '\\' + stuff
+                res = '%s[%s]' % (res, stuff)
+        elif c == '/':
+            # Special case of '/**/' which can match a single '/'
+            if i < n and rule[i] == '*' and rule[i - 1:i + 3] == '/**/':
+                i += 3
+                res = res + '.*/'
+            else:
+                res = res + '/'
+        else:
+            res += re.escape(c)
+    res += r'\Z'
+    return res
+
+
+class _Matcher:
+    """Couple a regex with other metadata for how we should match a given pattern."""
+
+    def __init__(self, line_num: int, orig_rule: str, invert: bool, only_dirs: bool,
```

Comment:
We're not using type annotations in the rest of the project (and in other parts of the file)... It bothers me to have these in only some places...

Reply:
It bothers me the internal inconsistency. But I guess that if they are needed in some points, it doesn't harm. What are you using them for?

Reply:
I'd rather have them than not. They are an easy way of documenting what's expected, and we should be able to grow them to the point where some static analysis can be done. I'm in no rush to get there, but I'd rather be closer than further away